### ISSUE TITLE
[Kind of RFE] Automatically switch to /run when /var becomes read-only?

### ISSUE THREAD

**Issue by grawity** [systemd/systemd#24410]
I'm using syslog-ng to forward messages from /var/log/journal to a network server. Whenever the system's rootfs becomes read-only (e.g. due to a disk issue on the VM host¹), syslog-ng no longer sees new messages so they immediately stop being forwarded. (And I had configured the log forwarding for the main purpose of receiving alerts about rootfs problems...)

[Despite /var/log becoming read-only, journald is still *kind of* able to log new messages, as they actually show up in a `journalctl -n 10` – but neither syslog-ng nor `journalctl --follow` notice them. I assume that's because journald is using mmap so any changes it makes still stay in the page cache (even if they can't be flushed to disk anymore), but of course it can no longer use its truncate() trick to notify readers about updates.]

If I manually run `journalctl --relinquish-var`, syslog-ng seems to actually notice the new .journal files in /run and starts pick up new messages like it should (and everything works if I just use Storage=volatile from the beginning, but that has its own downsides).

For now, I've configured that particular server to always use Storage=volatile (which hopefully also avoids tickling whatever issues the underlying host has), but I'm wondering whether journald should somehow automatically detect this situation and switch over to /run without being manually asked to.

<br>

¹ (The VM in question just routes IP packets – it happens to be hosted at a particularly useful location network-wise – and doesn't really store any data, so for now I'm not really worried about storage issues occurring in the first place.)

**Comment by poettering** [comment]
how would we notice though given that we use mmap writes? and how does one trigger this condition anyway? "mount -o remount,ro …" doesn't work...

**Comment by grawity** [comment]
> how would we notice though given that we use mmap writes?

I kind of remember that journald calls truncate() to trigger inotify events and wake up readers, so my first thought was it'd be returning -EROFS in this case as well? I didn't actually check via strace though if that happens.

> and how does one trigger this condition anyway? "mount -o remount,ro …" doesn't work...

I've used `echo u > /proc/sysrq-trigger` for testing this, it unmounts what it can (which is nothing in my case) but leaves rootfs remounted read-only.

(Also, I just remembered I can configure the legacy socket-based syslog message forwarding from journald to syslogd as an alternative, but in the end Storage=volatile ended up being a better choice in my case.)
